client/asset/btc/spv.go

```diff
@@ -0,0 +1,1712 @@
+		blockHash, limitHeight = w.mainchainBlockForStoredTx(txHash)
+		if blockHash == nil {
+			var err error
+			_, limitHeight, err = w.findBlockForTime(startTime)
+			if err != nil {
+				return nil, err
+			}
+		}
+	} else {
+		// No checkpoint, but user supplied a block hash.
+		var err error
+		limitHeight, err = w.getBlockHeight(blockHash)
+		if err != nil {
+			return nil, fmt.Errorf("error getting height for supplied block hash %s", blockHash)
+		}
+	}
+
+	w.log.Debugf("Performing cfilters scan for %v:%d from height %d", txHash, vout, limitHeight)
+
+	// Do a filter scan.
+	utxo, err := w.filterScanFromHeight(*txHash, vout, pkScript, limitHeight, checkPt)
+	if err != nil {
+		return nil, fmt.Errorf("filterScanFromHeight error: %w", err)
+	}
+	if utxo == nil {
+		return nil, asset.CoinNotFoundError
+	}
+
+	// If we found a block, let's store a reference in our local database so we
+	// can maybe bypass a long search next time.
+	if utxo.blockHash != nil {
+		w.storeTxBlock(*txHash, *utxo.blockHash)
+	}
+
+	w.cacheCheckpoint(txHash, vout, utxo)
+
+	return utxo, nil
+}
+
+// getTxOut finds an unspent transaction output and its number of confirmations.
+// To match the behavior of the RPC method, even if an output is found, if it's
+// known to be spent, no *wire.TxOut and no error will be returned.
+func (w *spvWallet) getTxOut(txHash *chainhash.Hash, vout uint32, pkScript []byte, startTime time.Time) (*wire.TxOut, uint32, error) {
+	// Check for a wallet transaction first
+	txDetails, err := w.wallet.walletTransaction(txHash)
+	var blockHash *chainhash.Hash
+	if err != nil && !errors.Is(err, WalletTransactionNotFound) {
+		return nil, 0, fmt.Errorf("walletTransaction error: %w", err)
+	}
+
+	if txDetails != nil {
+		spent, found := outputSpendStatus(txDetails, vout)
+		if found {
+			if spent {
+				return nil, 0, nil
+			}
+			if len(txDetails.MsgTx.TxOut) <= int(vout) {
+				return nil, 0, fmt.Errorf("wallet transaction %s doesn't have enough outputs for vout %d", txHash, vout)
+			}
+
+			var confs uint32
+			if txDetails.Block.Height > 0 {
+				tip, err := w.cl.BestBlock()
+				if err != nil {
+					return nil, 0, fmt.Errorf("BestBlock error: %v", err)
+				}
+				confs = uint32(confirms(txDetails.Block.Height, tip.Height))
+			}
+
+			msgTx := &txDetails.MsgTx
+			if len(msgTx.TxOut) <= int(vout) {
+				return nil, 0, fmt.Errorf("wallet transaction %s found, but not enough outputs for vout %d", txHash, vout)
+			}
+			return msgTx.TxOut[vout], confs, nil
+
+		}
+		if txDetails.Block.Hash != (chainhash.Hash{}) {
+			blockHash = &txDetails.Block.Hash
+		}
+	}
+
+	// We don't really know if it's spent, so we'll need to scan.
+	utxo, err := w.scanFilters(txHash, vout, pkScript, startTime, blockHash)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	if utxo == nil || utxo.spend != nil {
+		return nil, 0, nil
+	}
+
+	if utxo.blockHash == nil {
+		return nil, 0, fmt.Errorf("output %s:%v not found", txHash, vout)
```

Comment:
Instead of the usual `[TRC] CORE: Still searching for counterparty's contract coin ...`,
I'm getting `[ERR] CORE: Unexpected audit contract f89e23bc0...:0 (btc) error (will try again): error finding unspent contract: f89e23bc0...:0 : output f89e23bc0...:0 not found` for a couple dozen retries and then a hard fail:
`[ERR] CORE: Failed to audit contract coin %v (%s) for match %s: %v f89e23bc0...:0 btc ee22a089b7edd7280e885f14ff7688207cb3058fd49b5b0db331250009154a3b error finding unspent contract: f89e23bc0...:0 : output f89e23bc0....:0 not found`

I think this needs to return a `CoinNotFoundError` otherwise bad things happen in the caller.

Reply:
> Also, I'm surprised that since I'm trading with myself that walletTransaction didn't find the "counterparty" contract.... another proper testing challenge really here since a different btc wallet would be appropriate for a realistic test. Still, it must have fallen back to scanFilters, but I don't get why, even if that would be how trading with someone else would go.

We know the transaction, but we're not tracking the output's spend status, so we still have to scan.

> It also seemed to fail to find the tx with a cfilters scan in the block that contained the tx, and then start scanning the next (wrong) height.

Are you sure you're not seeing calls to `(*ExchangeWallet).SwapConfirmations` → `(*spvWallet).swapConfirmations` → `(*spvWallet).scanFilters`? Based on your logs, the swap was mined in 838 and Core found the swap on the first scan. `SwapConfirmations` would be called during `tick`s with the scan starting at 839`, which isn't mined yet, so the scan is a no-op because `limitHeight` > `tip`. And since there was a 2 conf requirement for btc, the swap couldn't proceed until 839 was actually mined. 

I'm not sure about the `It was a dcr tip change that triggered it`. That's the only part that doesn't make sense. 

> I think this needs to return a CoinNotFoundError

Oh. If the output isn't located, `getTxOut` should actually return a `nil` error, which has the same effect as `CoinNotFoundError` in `AuditContract`.

Reply:
Oh, is it only a 1-conf requirement for testnet btc?

Reply:
Ah. I see what was happening now. You're absolutely right. The `CoinNotFoundError` issue prevented us from advancing the match status because we hit the 24 `!errors.Is(err, asset.CoinNotFoundError)` limit, ending our `AuditContract` search early.